### PR TITLE
Commit to the extended data square

### PR DESCRIPTION
**Pull request by evan-forbes**
## Description

This PR generates the `DataHash` using erasured block data. Instead of creating a new method, like what is done using [`fillDataAvailabilityHeader`](https://github.com/celestiaorg/celestia-core/blob/2d3768eb5f6ea6bb75cd0893f48de461ece4280a/types/block.go#L119) in the current celestia-core master, this PR simply modifies the prexisting `Hash` method of `Data`. This approach has the benefit of being slightly closer to tendermint in that we have to change less lines of code, along with being a convenient way to not include the `DataAvailabilityHeader` inside of `Block`

Closes: #512
Part of #528 
blocked by #537 and #539

### CODE REVIEW

File: types/block.go [hunk 1]
```diff
@@ -1104,12 +1038,26 @@ type Data struct {
...
-		return (Txs{}).Hash()
+	if data.hash != nil {
+		return data.hash
 	}
-	if data.hash == nil {
-		data.hash = data.Txs.Hash() // NOTE: leaves of merkle tree are TxIDs
+
+	// compute the data availability header
+	// todo(evan): add the non redundant shares back into the header
+	shares, _ := data.ComputeShares()
```
Comment (evan-forbes) [inline types/block.go:1047]: this is personal note to properly write an issue to handle the second returned value from `ComputeShares` and include it in the header. I think it should be handled in a different PR tho, as it an optimization. That being said, if anyone thinks we should handle this now, we can do that too :slightly_smiling_face:

File: types/block.go [hunk 2]
```diff
@@ -1104,12 +1038,26 @@ type Data struct {
...
+	rawShares := shares.RawShares()
+
+	squareSize := uint64(math.Sqrt(float64(len(shares))))
+
+	eds, err := da.ExtendShares(squareSize, rawShares)
+	if err != nil {
+		panic(err)
 	}
```
Comment (evan-forbes) [inline types/block.go:1055]: we panicked before in [`fillDataAvailabilityHeader`](https://github.com/celestiaorg/celestia-core/blob/2d3768eb5f6ea6bb75cd0893f48de461ece4280a/types/block.go#L127), but I think this time it might be worse, as `ExtendShares` will return an error if the square size is too large!

**Comment by evan-forbes** [comment]
All the tests are passing for me locally, and the CI is showing some weird errors on all the "failing" tests.
```
Error: command [git diff] exited with code 128. message: fatal: ambiguous argument 'get-diff-action/evan/cherry-pick-share-merging-splitting...get-diff-action/pull/540/merge': unknown revision or path not in the working tree.```

**Comment by liamsi** [comment]
Sounds like this is a github-action only problem? Maybe it's just caching? Did you try restarting the process manually?

**Comment by liamsi** [comment]
What I would try to make sure this isn't some super weird cache issue, I'd create a new branch and see if the error still happens there. This issue sounds related: https://github.com/technote-space/get-diff-action/issues/164

**Comment by evan-forbes** [comment]
Thanks for helping me debug, I tried to debug quite a few times, not sure why it started working now. I'll merge this as soon as the CI passes.
